web/src/components/molecules/Asset/AssetList/index.tsx

```diff
@@ -42,32 +45,32 @@ const AssetList: React.FC<Props> = ({
   uploading,
   uploadModalVisibility,
   loading,
-  onAssetCreate,
+  uploadUrl,
+  uploadType,
+  onUploadModalCancel,
+  setUploadUrl,
+  setUploadType,
+  onAssetsCreate,
+  onAssetCreateFromUrl,
   onAssetDelete,
   onSearchTerm,
   onEdit,
   setSelection,
   setFileList,
-  setUploading,
   setUploadModalVisibility,
   onAssetsReload,
 }) => {
   const displayUploadModal = useCallback(() => {
     setUploadModalVisibility(true);
   }, [setUploadModalVisibility]);
 
-  const hideUploadModal = useCallback(() => {
-    setUploadModalVisibility(false);
-    setUploading(false);
-    setFileList([]);
-  }, [setFileList, setUploading, setUploadModalVisibility]);
-
-  const handleUpload = useCallback(() => {
-    setUploading(true);
-    onAssetCreate(fileList).finally(() => {
-      hideUploadModal();
-    });
-  }, [fileList, setUploading, onAssetCreate, hideUploadModal]);
+  const handleUpload = useCallback(async () => {
+    if (uploadType === "url") {
+      await onAssetCreateFromUrl(uploadUrl);
+    } else {
+      await onAssetsCreate(fileList);
+    }
+  }, [uploadType, onAssetCreateFromUrl, uploadUrl, onAssetsCreate, fileList]);
```

Comment:
If you are using `onAssetCreateFromUrl` and `onAssetsCreate` only in this callback, can you move the callback up to where the `uploadUrl` is declared and then pass just `onUpload={handleUpload}`?

Reply:
Also, does this work correctly? I see you are using handleUpload like `const asset = handleUpload()` Do you need to do something like this to have something returned?
```
const result = await onAssetCreateFromUrl(uploadUrl);
return result
```

Reply:
For `AssetList` page I use this function and I do not need to return anything since there is no linking involved from there.
https://github.com/reearth/reearth-cms/blob/a1c09d2f850f0b257e98aa7f219d6573dce8fea1/web/src/components/molecules/Asset/AssetList/index.tsx#L67

However, for `LinkAssetModal` that is used in `Schema` and `Content` pages I use a different function called `handleUploadAndLink`
https://github.com/reearth/reearth-cms/blob/a1c09d2f850f0b257e98aa7f219d6573dce8fea1/web/src/components/molecules/Common/Form/AssetItem/hooks.ts#L38

Both of these function are passed to the same `UploadAsset` component.